### ISSUE TITLE
Fish versioning broken under cmake

### ISSUE THREAD

**Issue by mqudsi** [fish-shell/fish-shell#4626]
Tested by building with cmake with both the ninja and regular Make generators, `fish --version` and `echo $version` are both broken in `master`.

**Comment by mqudsi** [comment]
This doesn't seem right, but I checked out 9f8ae95b0e201b13a6cdd84b2c1436999377cd81 and built it (had to specify `-DCMAKE_CXX_STANDARD_LIBRARIES='-ldl -pthread'`) and they were still broken. The build script does print the expected `FISH_BUILD_VERSION=2.7.0-604-g87f7cd0` output on running `cmake ..`

**Comment by zanchey** [comment]
Yeah, the problem is that `fish_version.cpp` expects the FISH_BUILD_VERSION define to be present, but the CMake build doesn't do that yet.

This is the relevant code from the Makefile:
```
FISH-BUILD-VERSION-FILE: FORCE | show-FISH_BUILD_VERSION
        $v build_tools/git_version_gen.sh 2> /dev/null
-include FISH-BUILD-VERSION-FILE
CPPFLAGS += -DFISH_BUILD_VERSION=\"$(FISH_BUILD_VERSION)\"
.PHONY: FORCE
obj/fish_version.o: FISH-BUILD-VERSION-FILE
```

**Comment by mqudsi** [comment]
OK, I updated the CMake build to inject the needed info. It's more straightforward with cmake than it was with the old Makefile script.

@zanchey and @ridiculousfish please note that I purposely made a copy of the versioning script rather than modify the existing so I could both take advantage of cmake's idiomatic approach for version injection and prevent the old build script from breaking.

**Comment by mqudsi** [comment]
Oh, in case it's not obvious, cmake builds are deterministic in that each time `ninja` or `make` is run, the output isn't supposed to change, so for the git version info to be correctly updated `cmake ..` will first need to be executed.

**Comment by mqudsi** [comment]
I don’t like this. If someone forgets to rerun cmake but has changed the code, it is wrong for the version number to not reflect that and I can see this causing problems down the line. The versioning code should be run regardless of whether the project is reconfigured.

**Comment by anordal** [comment]
That wouldn't be a problem if we used Meson.

[Do I ever have to rerun Meson?](https://groups.google.com/forum/#!topic/mesonbuild/FFOD92YNH2Q)

> You only ever need to run the Meson binary manually when setting up a build directory for the first time. After that you always run Ninja. No matter what changes you do, just run Ninja, it will sort everything out. If it doesn't, that is an error and you should file a bug.

**Comment by zanchey** [comment]
@mqudsi, yes - a deliberate goal of the earlier implementation was to avoid exactly this case.

**Comment by mqudsi** [comment]
@anordal I think everyone's preference at the moment is to get cmake working. Also, https://github.com/mesonbuild/meson/issues/2335

**Comment by ridiculousfish** [comment]
@anordal both CMake and Meson will re-invoke the meta-build system if the meta-build file timestamps have changed. But "the output of git-describe" doesn't fit nicely into the timestamp detection machinery, hence the issue. I don't think Meson solves this problem any better (if it does I'd like to know how!)

**Comment by ridiculousfish** [comment]
I have this almost working nicely [in a branch](https://github.com/ridiculousfish/fish-shell/commits/fix-versioning). The main blocker is actually make (not Ninja) - make is still building too much.

**Comment by ridiculousfish** [comment]
I think this should have been addressed once and for all by d09210c08bc8e36ac32727ef12bf3ce35f5374ee.

**Comment by anordal** [comment]
> I don't think Meson solves this problem any better (if it does I'd like to know how!)

After seeing how Mesa did it, I created this toy example:

[Meson: Unconditionally generate source file](https://gist.github.com/anordal/c34535b581355b28d4f4b58d71efe4d8)
